### PR TITLE
Use bloom filters to limit what servers account deletion notices are sent to

### PR DESCRIPTION
**Pull request by ClearlyClaire**
One outstanding issue with Mastodon is that deleting a local account sends a `Delete` activity to the whole known fediverse to ensure everyone is aware that the account does not exist anymore. This is wasteful, and increasingly so with the growth of the fediverse. It is also a minor privacy concern, as servers who would otherwise not know about an account would learn about its (previous) existence without a good reason.

Unfortunately, tracking which servers are aware of an account is costly and difficult to do, and fundamentally incompatible with reverse-proxy caching. However, some servers chose to run with `AUTHORIZED_FETCH=true`, which disables caching and requires a valid signature for fetching any account. This means that in this configuration, Mastodon can be aware of all the servers which know about a particular account.

This PR introduces the `bloom_filters` experimental feature that enables a relaxed version of `AUTHORIZED_FETCH=true` that still allows anonymous access to public *posts* and still signs them for efficient forwarding, but requires valid signatures for account fetches. This means it has a lot of the performance concerns of `AUTHORIZED_FETCH=true` but not its usability concerns.

The PR also uses bloom filters to keep track of which domains have fetched each account, and only sends the `Delete` activities to servers whose domain matches the bloom filter. The better-known the account is, the less accurate its filter will be. Nonetheless, this should still make a huge difference for short-lived accounts and reasonably small ones.

Note that this will only work for accounts created after this code is in effect, and only as long as they are created and only fetched when `AUTHORIZED_FETCH` is set to `actors` or `true`.

## TODO / Concerns

- ~~currently, the bloom filter update is prone to race conditions~~ that should be fixed with the use of a lock in the worker
- ~~it is also fairly inefficient, with a database round-trip of the whole filter for each update~~ that may be fixed by the use of a Redis set and worker with `until_executing`
- ~~we probably want to review the bloom filter's hash functions and make it so that an attacker cannot easily construct a set of subdomains that would maximize collisions (e.g. by using some salt)~~ should be addressed by a per-filter salt
- are there more efficient ways to go about it than bloom filters?
- we should evaluate the performance impact of `AUTHORIZED_FETCH=actors` and consider making it the default

### CODE REVIEW

**Comment by dariusk** [comment]
I really like the intent behind this. I don't know anything about bloom filters myself (though I'll be reading up today, at first glance they seem interesting). But I'm interested to know more about "the better known the account is, the less accurate its filter will be." Do I have the correct sense of "less accurate" where the end result is that there will be some number of servers M of all known servers N that do not receive the Delete activity? Is there an upper bound to M that is lower than N, or does inaccuracy approach 100% as the number of remote servers tracked increases (I imagine not but it's worth asking)?

Basically I'm trying to understand how bad the inaccuracy gets as an account is known by more remote servers.

**Comment by ClearlyClaire** [comment]
> Do I have the correct sense of "less accurate" where the end result is that there will be some number of servers M of all known servers N that do not receive the Delete activity?

No, it's that the more unique domains are registered in the filter, the more false positives there will be, that is, known servers that did not previously know about the account will be sent the `Delete` activity. A 100% error rate means all known servers will be sent the `Delete` activity, even if they did not ever fetch the actor (that is, the current behavior).

> Is there an upper bound to M that is lower than N, or does inaccuracy approach 100% as the number of remote servers tracked increases (I imagine not but it's worth asking)?

The inaccuracy does approach 100% as the number of remote servers tracked increases. The parameters are such that under 10k tracked servers, the error rate should be lower than 10% (and the less tracked servers, the less inaccuracy), at the cost of a ~6k bloom filter per local account.

This is difficult to evaluate because there is no real bound on the number of servers that will be known, nor on the number of servers that will know each account.

That being said, one of the largest Mastodon instances, `mastodon.social`, currently knows about 44k unique domains, and my almost 6 years-old single-user instance knows about 10k of them.

I am unsure how relevant those numbers are, but:
- an account known by all of the 10085 servers known to my instance would match 13565 of the 44k servers known to mastodon.social, that's still over 3k of wasted requests, but over 30k of spared request compared to the current situation
- I *believe* an account that interacts with Gargron would currently be known to around 3k servers. Creating a filter based on 3k random servers known by mastodon.social yields 3105 matching servers, which is 105 wasted requests, but over 40k saved queries compared to the current situation.
- an account known by only 600 random servers picked from mastodon.social's list of known servers would match 601 of them, so only 1 wasteful query

File: app/controllers/accounts_controller.rb [hunk 1]
```diff
@@ -32,13 +33,39 @@ def show
...
     end
   end
 
   private
 
+  def record_reach!
+    unless authorized_fetch_mode?
+      AccountReachFilter.where(account_id: @account.id).delete_all
+      return
+    end
```
Comment (ThisIsMissEm) [inline app/controllers/accounts_controller.rb:50]: Is this so if you turn off authorized fetch mode all the bloom filters are deleted? This seems like it should be a CLI task?

File: app/controllers/accounts_controller.rb [hunk 2]
```diff
@@ -32,13 +33,39 @@ def show
...
+
+    # It's not ideal, but we have no way to tell the remote account about it…
+    return if signed_request_account.preferred_inbox_url.blank?
+
+    account_reach_filter_id = AccountReachFilter.where(account_id: @account.id).pick(:id)
+    return if account_reach_filter_id.nil?
+
+    with_redis do |redis|
+      redis.sadd("account_reach:#{account_reach_filter_id}:to_add", Addressable::URI.parse(signed_request_account.preferred_inbox_url).normalized_host)
+    end
```
[the remark this answers is not in the document]
Comment (ThisIsMissEm) [reply app/controllers/accounts_controller.rb:60]: That or sending the `inbox_url` via an argument to `UpdateAccountReachWorker.perform_async`? Just feels a bit odd to be managing data directly in redis here.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -51,7 +51,7 @@ def raise_not_found
   private
 
   def authorized_fetch_mode?
-    ENV['AUTHORIZED_FETCH'] == 'true' || Rails.configuration.x.whitelist_mode
+    %w(true all).include?(ENV.fetch('AUTHORIZED_FETCH', 'false')) || Rails.configuration.x.whitelist_mode
```
Comment (ThisIsMissEm) [inline app/controllers/application_controller.rb:54]: Should this read?
```suggestion
    %w(true all actors).include?(ENV.fetch('AUTHORIZED_FETCH', 'false')) || Rails.configuration.x.whitelist_mode
```

**Review by ThisIsMissEm — COMMENTED** [review]
This looks okay, but I've definitely some questions about the code / where implementation details seem to be maybe leaking.

File: app/controllers/accounts_controller.rb (hunk 1 above)
Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb:50]: That's it. I'm not sure this should be a CLI task, not deleting the bloom filter opens you to miss recipients in your delivery of `Delete` activities.

File: app/controllers/accounts_controller.rb [hunk 4]
```diff
@@ -32,13 +33,39 @@ def show
...
+    unless authorized_fetch_mode?
+      AccountReachFilter.where(account_id: @account.id).delete_all
+      return
+    end
+
+    # It's not ideal, but we have no way to tell the remote account about it…
+    return if signed_request_account.preferred_inbox_url.blank?
+
+    account_reach_filter_id = AccountReachFilter.where(account_id: @account.id).pick(:id)
+    return if account_reach_filter_id.nil?
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb:56]: Existing accounts don't have a reach filter: if we created reach filters for existing accounts, they would be incomplete, missing all the servers that did know of that account before the feature got implemented/turned on.

File: app/controllers/accounts_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb:60]: `UpdateAccountReachWorker` works in batches, and will not be queued multiple times, hence why we don't pass it as an argument. Moving this to an `AccountReachFilter#enqueue_addition` makes sense though.

File: app/controllers/accounts_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (ThisIsMissEm) [reply app/controllers/accounts_controller.rb:56]: At the very least then, this should be checked and done on application startup, not at request time.

File: app/controllers/accounts_controller.rb (hunk 1 above)
Comment (ThisIsMissEm) [reply app/controllers/accounts_controller.rb:50]: At the very least then, this should be checked and done on application startup, not at request time.

File: app/controllers/accounts_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (ThisIsMissEm) [reply app/controllers/accounts_controller.rb:56]: Couldn't we backfill this based on the follower graph, and perhaps also boosts & replies?

[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb:56]: This would be very costly, and still incomplete.

File: Gemfile.lock [hunk 5]
```diff
@@ -151,6 +151,7 @@ GEM
     bindata (2.4.15)
     binding_of_caller (1.0.0)
       debug_inspector (>= 0.0.1)
+    bloom-filter (0.2.2)
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply Gemfile.lock:154]: I think I had evaluated it too, but I honestly can't remember why I decided not to go with it.

**Comment by mjankowski** [comment]
This one needs rebase and CI approval.

Separately -- is this sufficiently far along (and desired, likely to merge) that it could use code review at this point?

**Comment by mjankowski** [comment]
Is this still relevant/WIP, or abandoned?

If relevant, probably needs rebase and CI re-trigger. Let me know if code review is useful in current state.

File: app/controllers/accounts_controller.rb [hunk 6]
```diff
@@ -6,10 +6,11 @@ class AccountsController < ApplicationController
 
   include AccountControllerConcern
   include SignatureAuthentication
+  include Redisable
 
-  vary_by -> { public_fetch_mode? ? 'Accept, Accept-Language, Cookie' : 'Accept, Accept-Language, Cookie, Signature' }
+  vary_by -> { authorized_fetch_actors? ? 'Accept, Accept-Language, Cookie, Signature' : 'Accept, Accept-Language, Cookie' }
```
Comment (renchap) [inline app/controllers/accounts_controller.rb]: We should not use `Vary` with a header that have a high cardinality. Better to not cache those requests at all

Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb]: The requests do not get cached when a `Signature` is passed, it does get cached when there is no `Signature`. This is the way we do it in other places as well. It just lets the reverse-proxy know to not serve the Signature-less cached version to requests with a Signature.

File: app/controllers/accounts_controller.rb (hunk 1 above)
Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb]: Kept the behavior as-is because we may want to keep actors that are not re-queried as is. Still a bit wasteful though, but not that much.

File: app/controllers/accounts_controller.rb [hunk 7]
```diff
@@ -32,13 +33,39 @@ def show
...
+
+    with_redis do |redis|
+      redis.sadd("account_reach:#{account_reach_filter_id}:to_add", Addressable::URI.parse(signed_request_account.preferred_inbox_url).normalized_host)
+    end
+
+    UpdateAccountReachWorker.perform_async(account_reach_filter_id)
+  end
+
+  def authorized_fetch_actors?
+    %w(true all actors).include?(ENV.fetch('AUTHORIZED_FETCH', 'false')) || Rails.configuration.x.whitelist_mode
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb]: Can't be used in the Rails configuration layer because it can also be set through an admin setting, but it's all abstracted in `AuthorizedFetchHelper` now.

File: app/controllers/accounts_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/accounts_controller.rb]: Moved to `AccountReachFilter.record_reach_for`. Not an instance method because we want to avoid loading it from the database (I suppose we could lazy-load attributes, but that seems error-prone).

File: Gemfile.lock (hunk 5 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply Gemfile.lock]: Using `bloom_fit` instead.

File: app/services/delete_account_service.rb [hunk 8]
```diff
@@ -275,7 +275,15 @@ def delivery_inboxes
   end
 
   def low_priority_delivery_inboxes
-    Account.inboxes - delivery_inboxes
+    inboxes = Account.inboxes - delivery_inboxes
+    if @account.reach_filter.present?
+      inboxes.filter! do |url|
+        @account.reach_filter.include?(Addressable::URI.parse(url).normalized_host)
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/services/delete_account_service.rb]: Moved logic to `AccountReachFilter.record_reach_for` and `AccountReachFilter#filter_inboxes`, so this implementation detail is limited to `AccountReachFilter` now.
